### ISSUE TITLE
Feature request: Contractual middleware

### ISSUE THREAD

**Issue by ryanstaniforth** [ts-rest/ts-rest#357]
I would like to be able to define middleware in the contract itself and have it abstracted away from the server implementation (express/fastify).

## Problem

I'll use authentication as my example, as I think this is the most common use of middleware. If I want to implement auth currently, I have a few options:

### Do the auth check inside the handler and throw

- ❌ Leads to repeated code in each handler.
- ❌ Using a thrown error to send the appropriate response does not adhere to the idea of a contract. (We can't know how the thrown error is handled and what response status/body is sent.)

```typescript
getSelf: async ({ request }) => {
  const authenticateduser = getAuthenticatedUserOrThrow(request)

  return {
    status: 200,
    body: authenticateduser
  };
}
```

### Do the auth check inside the handler and return correct response from inside handler 

- ✅ 401 response is understood by the contract.
- ❌ Even more repeated code.

```typescript
getSelf: async ({ request }) => {
  const authenticateduser = getAuthenticatedUser(request)

  if (!authenticateduser) {
    return {
      status: 401,
      body: {
        message: 'Unauthenticated'
      }
    }
  }

  return {
    status: 200,
    body: authenticateduser
  };
}
```

### Use contract metadata to specify that a route requires authentication

- ✅ The `request` object can be modified to contain the authenticated user by the server when metadata tells it to.
- ❌The authenticated user in the `request` object would have to be optional.
- ❌ The 401 response on auth failures is not understood by the contract.

## Proposal

Provide a `Middleware` type. Each middleware defines the responses it can give and its handler. The handler is passed objects similar to a route handler such as `request`, `headers` and `body`. (`headers` and `body` will be more loosely typed, as middleware will be generic to any route).

```typescript
export const authenticate: Middleware<{ authenticatedUser: User }> = {
  responses: {
    401: z.object({
      message: z.string(),
    })
  },
  handler: async ({ request, body, headers }) => {
    // auth logic.

    if (!user) {
      return {
        status: 401,
        body: {
          message: 'Unauthenticated'
        }
      }
    }
  
    return {
      authenticatedUser: user
    }
  }
}
```

Allow the contract to accept a list of middleware. The route understands the responses given by the middleware and can therefor merge them with its specific `responses`.

```typescript
const selfContract = c.router({
  getSelf: {
    method: 'GET',
    path: '/self',
    middleware: [authenticate],
    responses: {
      200: Self,
      401: UnauthenticatedResponse,
    },
  },
});
```

The route implementation could accept another object that combines all response objects from each middleware it has.

```typescript
export const selfRouter = s.router(contract.self, {
  getSelf: async ({ request, middleware: { authenticatedUser } }) => {
    return {
      status: 200,
      body: authenticateduser
    };
  }
}
```

- ✅ Middleware responses are typed within the contract
- ✅ Middleware responses are not repeated on each route
- ✅ No repeated logic inside route handlers. Data such as `authenticatedUser` passed to the route handler does not need to be checked for existence as we know that we wouldn't make it to the handler if auth failed.

**Comment by oliverbutler** [comment]
This is super interesting proposal, definitely feels very tRPC-y

My initial gut reaction is worrying about bundling server side code into the client. If this is a shared library in a larger repo it would now have to import your database/services/etc into the contract.

The client side bundler would then hopefully be smart enough to not include backend code. 

Interested to hear other thoughts here, as I definitely agree with the problem you're trying to solve.

**Comment by ryanstaniforth** [comment]
@oliverbutler Completely agree about not bundling server-side code with client. I realised today that the middleware would also need to be defined firstly in the contract and then implemented separately on the server, just like we do with routes.

I think middleware is differs from routes in the sense that it's reusable across multiple selected routes. Because of this, I think it needs defining separate from the route contract(s), and using its own method/helper for doing so.

```typescript
// Create middleware contract
const authenticateMiddlewareContract = c.middleware({
  // We need a way of referencing the middleware during run-time.
  id: 'authenticate',
  // The responses the middleware can conditionally return. If one of these is returned, the route handler is not executed.
  responses: {
    401: z.object({
      message: z.string(),
    })
  },
  // This is the data that will conditionally be sent to the route handler.
  data: z.object({
    authenticatedUser: User
  })
});
```

Here we can pass it to the route contract. This will provide type information when implementing the route server-side, allowing the handler to know the object shape passed from the middleware (`{ authenticatedUser: User }` in this case)
```typescript
const selfContract = c.contract({
  getSelf: {
    method: 'GET',
    path: '/self',
    // Use the middleware in the route contract
    middleware: [authenticateMiddlewareContract],
    responses: {
      200: Self,
      401: UnauthenticatedResponse,
    },
  },
});
```

We need to implement each middleware and register them somewhere. As we're not defining the middleware in with the route contract, I think doing this at `s.plugin()` makes sense. This is probably the part that's most up for debate.

```typescript
s.plugin(router, [
  {
    id: 'authenticate',
    handler: async ({ request }) => {
      // authentication logic.

      if (authenticated) {
        // return data
        return { data: user };
      } else {
        // or one of the defined responses.
        return { response: { status: 401, body: { message: 'Unauthenticated' } } }
      }
    }
  }
])
```

```typescript
// When it comes to implementing the server-side route handler, the middleware setup is already complete.
export const selfRouter = s.router(contract.self, {
  getSelf: async ({ request, middleware: { authenticatedUser } }) => {
    // If this handler is executed, the middleware must have passed data (rather than a response).
    return {
      status: 200,
      body: authenticateduser
    };
  }
}
```

Hope I've made sense with all that 😄

**Comment by ryanstaniforth** [comment]
Small update: I think the middleware needs the ability to *add* headers without returning a response (i.e. allowing the request to reach the route handler).

Useful for creating CORS middleware for example:

```typescript
return { data: null, headers: { 'Access-Control-Allow-Origin': '*' } };
```

The route handler can still add headers to the response. They will just be merged with any from the middleware.

**Comment by llllvvuu** [comment]
> Completely agree about not bundling server-side code with client. I realised today that the middleware would also need to be defined firstly in the contract and then implemented separately on the server, just like we do with routes.

I like this idea of a schema-only (no implementation) middleware definition.

> I think middleware differs from routes in the sense that it's reusable across multiple selected routes.

I wonder if handling arbitrary routes may require defining as a schema-to-schema function rather than a static schema:

```typescript
type MiddlewareSchema<TReqSchema extends AnyZodObject, TResSchema extends AnyZodObject>
  = (reqSchema: TReqSchema, resSchema: TResSchema) => AnyZodObject
```

This would work really well at runtime, although it loses compile-time information. One would need type-level computation to make the return type not just `AnyZodObject` but a type-level function of `TReqSchema` and `TResSchema`.

This black magic is beyond my depth, zod's type-inference might just be able to do it, if the generic is updated to delegate to `ZodObject` in the parameter signature (here I'm copy-pasting from `zod/lib/types.d.ts`):
```typescript
type MiddlewareSchema<
  TReq extends ZodRawShape,
  ReqUnknownKeys extends UnknownKeysParam,
  ...,
  TRes extends ZodRawShape,
  ReqsUnknownKeys extends UnknownKeysParam,
  ...
> = (reqSchema: ZodObject<TReq, ReqUnknownKeys, ...>, resSchema: ZodObject<TReq, ResUnknownKeys, ...>) => /** type inference */
```

A limitation of the above approach is that it is even more coupled to `zod`, although maybe `ts-rest` is already maximally coupled to zod.

**Comment by llllvvuu** [comment]
tl;dr the few ideas above:
1. `(req, res) -> res` middleware schema with type inference support
2. `(req, res) -> res` middleware schema without type inference support
3. `extendBy` (i.e. subset of middleware schemas) with type inference support
4.
    a. Top-level `responses` / (optional) `query` / (optional) `requestHeaders`, with type inference (post `z.infer`) so n need to mess with `zod ` TypeScript types) support
    b. User needs decorator/higher-order function to make handlers return the desired type. Handler types are post-`z.infer` so easier to work with.

3. and 4. are scoped down from supporting arbitrary middleware schema

**Comment by bnsngltn** [comment]
Looking forward to this as well!

As a temporary workaround, I've created [this gist](https://gist.github.com/bnsngltn/0ac0236e38acadcb22fbe5ed071f3964) on how to integrate `next-auth` and contracts together. You can swap out  `next-auth` with any auth provider/workflow. 

The usage is something like this:

```typescript
import { protectedHandler, publicHandler } from "@/server/ctx";
import { contract } from "./contract"

const postRouter = createNextRoute(contract, {
  public: publicHandler(async (args) => {
     // session is nullable
     args.ctx.session?.user
  }),
  protected: protectedHandler(async (args) => {
    // session and user is always defined 
    args.ctx.session.user
  }),
});
```

The `args.body` and the return type for each handler (route) is still fully inferred from the contract.

It's heavily inspired from `tRPC` though. The only downside of this approach for now is that it's not fully reflected on the contract that a route requires authentication to access.
